data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.expression.util;
+
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.hamcrest.Description;
+import org.hamcrest.DiagnosingMatcher;
+import org.hamcrest.Matcher;
+
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.opensearch.dataprepper.expression.util.ContextMatcher.describeContextTo;
+
+/**
+ * Base class for Matcher that asserts a unary tree ending in a node that matches {@link SimpleExpressionMatcher#lastNodeClass}
+ */
+abstract class SimpleExpressionMatcher extends DiagnosingMatcher<ParseTree> {
+    private static final Matcher<Integer> childCountMatcher = is(1);
+
+    protected final List<Class<? extends ParseTree>> validRuleOrder;
+    protected final Class<? extends ParseTree> lastNodeClass;
+
+    protected SimpleExpressionMatcher(final List<Class<? extends ParseTree>> validRuleOrder) {
+        this.validRuleOrder = validRuleOrder;
```

Comment:
Can we also store List of validRuleTypeOrder which store the type ? The getRuleIndex method will be reduced to below (you also wouldn't require this private method):

```
private int getRuleIndex(final ParseTree item) {
 return validRuleTypeOrder.indexOf()
}
```

Reply:
Thanks for the idea @dinujoh. I'm not sure I completely understand your suggestion, in an earlier iteration of `getRuleIndex` I used `indexOf()`. Cases with subclasses did not work as expected.

A quick example:

```java
    private class NewContext extends DataPrepperExpressionParser.ExpressionContext {
        public NewContext(final ParserRuleContext parent, final int invokingState) {
            super(parent, invokingState);
        }
    }
    
    private int getIndex() {
        final List<Class<? extends ParseTree>> validRuleOrder = Arrays.asList(
                DataPrepperExpressionParser.ExpressionContext,
                DataPrepperExpressionParser.ConditionalExpressionContext,
                DataPrepperExpressionParser.RegexOperatorExpressionContext
        );
        
        validRuleOrder.indexOf(NewContext.class); // expected 0, will return -1
    }
```

Reply:
For the subclass use case to work, you need to override the indexOf method:

```
List<Class<? extends ParseTree>> _validRuleOrder = new ArrayList<Class<? extends ParseTree>>(validRuleOrder) {
    @Override
    public int indexOf(Object item) {
        for (int x = 0; x < this.size(); x++) {
           if (this.get(x).isInstance(item)) {
              return x;
           }
        }
       return -1;
    }
};
```

Looks like validRuleOrder List data structure is used to find the index of the type/subtype of passed in item. We can define another data structure and abstract the operations. SimpleExpressionMatcher will be much cleaner and maintainable. 

```
public class RuleOrderTypeMatcher <T> {
  private final List<Class<? extends T>> ruleOrderByType;
  public RuleOrderTypeMatcher(final List<Class<? extends T>> validRuleOrder) { 
      List<Class<? extends T>> ruleOrderByType = new ArrayList<Class<? extends T>>(validRuleOrder) {
    @Override
    public int indexOf(Object item) {
        for (int x = 0; x < this.size(); x++) {
           if (this.get(x).isInstance(item)) {
              return x;
           }
        }
       return -1;
    }
   };
  }

 public boolean isValidRuleOrder(..) {
 ...
 }

 public boolean matchesLastNode(final T item) {
 ..
 }
}
```

Reply:
Thanks for making this change.